### ISSUE TITLE
Change of  learning rate  when using scheduler during training? 

### ISSUE THREAD

**Issue by ranchlai** [espnet/espnet#3805]
It is possible to change learning rate during training when using a Scheduler? 
I see that Scheduler state will be loaded when training resumes, so changing learning is not supported? Any plan to make training more flexible?

**Comment by kamo-naoyuki** [comment]
Do you mean, changing the learning rate manually instead of the scheduled learning rate?

1. Start training
2. Stop training
3. Manual operation for hyper parameters
4. Resume

If you really want to do it, please modify the optimizer state directly from the saved file. It's easy, but why don't you use a custom scheduler? I think it's enough.

**Comment by ranchlai** [comment]
I think it is much easier to modify config.yaml than to modify checkpoint, and it is scary  and troublesome for most people to load checkpoint, modify it and save it back.  May be we can change the checkpointing features to enable using new scheduler parameters?

**Comment by kamo-naoyuki** [comment]
Learning rates are saved in the scheduler state.  I have no idea to enable it. Maybe, it's impossible.

**Comment by kamo-naoyuki** [comment]
Anyway, modifying config.yaml is not recommended as well as checkpoint. Samely troublesome. 

We never provide such a feature officially.

**Comment by ranchlai** [comment]
Changing hyper-parameters during training seems to be very common, especially in research/competition. How about checking diff between checkpoint and config.yaml, and if possible, use params from config to update the checkpoint? I did that in my private fork and I think it is useful.

**Comment by kamo-naoyuki** [comment]
Hmm. I never did it. How  do you change hyperparameters? By checking loss curve? If so, why don't you scheduler?

**Comment by ranchlai** [comment]
For LR, sometimes we need to manually set the LR to see how the loss goes. Even the scheduler has its own parameters that we need to take care of. But this is not just the for LR. For ASR, there are other hyper-params such as spect-augment, rir, ctc-weights and so on.  If changing these params requires training from scratch, it will cost a lot of energy.

**Comment by kamo-naoyuki** [comment]
Okay, I'll consider it (Though I believe such people should modify the checkpoint file directly because it's a rare case for reseachers)

At the first, trainer can't get the inforamation of the configuration directly because trainer is indepedent from the configuration in the current design. Thus we need to give the information from AbsTask to trainer.

```python
# In abs_task.py
trainer.run(
  ...,
  is_optimizer_conf_changed=check_optim_conf(...),
  is_scheduler_conf_changed=check_scheduler_conf(...),
)
```

Next, to change hyperparameters after instantiation of optimizer/scheduler, we need to set it by accessing to each attribute directly or changing the state information, but this is not recommended in general.  Actually, this is impossible for general optimizers because we can't know what parameters exist in advance.

```
optim.param_groups[0]["lr"] = ...
``` 

Thus, I think only we can do is  not loading optimzer state / scheduler state when the configuration is changed.

```python
# In trainer.py
if is_optimizer_conf_changed:
    pass
elif  is_scheduler_conf_changed:
  load_optimzer()
else:
  load_optimzer()
  load_scheduler()
```

**Comment by kamo-naoyuki** [comment]
Does it work for you?

**Comment by kamo-naoyuki** [comment]
Hmm. I thought my idea is a bad design because someone may want to load the optimizer state even if changing the configuration, e.g. the changed parameter is not saved in the state. We can't know what parameters are saved in the state in general.

How about adding such options as `--disable_load_optimizer_state` and `--disable_load_scheduler_state`?

**Comment by kamo-naoyuki** [comment]
Maybe, with no change, it's enough to launch a new training process with `--init_param`?  I think this is a better way for reproducibility. How about it?

**Comment by kamo-naoyuki** [comment]
Another idea is adding an option to set attributes directly for optimizer, such as `--overwrite_optim1_param`

```python
optim1.param_groups[0].update(overwrite_optim1_param)
```

In this way, we can set attributes directly because we don't need to know the parameter names in advance.

**Comment by kamo-naoyuki** [comment]
I think there are two cases to modify lr in practice:

- Reducing lr with keeping the lr curve ( I mean keeping the state of scheduler with reducing lr relatively)
- Reseting the scheduler and optimizer and setting new hyper parameters.

The second can be enabled by `--init_param`. I think this is the best way for the second case.

The behaviour of `--disable_load_*_state` differs depending on the scheduler implementation, e.g. for `ReduceLROnPlateau`, it performs as the second type, while it performs as the first type for `CosineAnnealingLR` because the lr curve is decided only by `epoch`. At the worst case, it is possible to create unexpected curve if it depends on `epoch` and the state of scheduler. 
I think this option may be  unexpected, so I decided not to implement it.

PyTorch doesn't provide the interface for changing parameters of scheduler safely, so I have no idea to enable the first case for all hyper parameters, but if restricting only to changing lr, it is possible without lossing generality. 

As your idea, detecting the changing of lr in configuration is also possible, but this is impossible for the other hyper parameters. It is confusing if we enable this feature only for lr, so I decided not to implement it.

Finally, my conclusion is adding `--overwrite_optim{num}_lr`: This option overwrites both the lr of optimizers and `base_lrs` for the inheritance of `_LRScheduler` when resuming .

Is it okay for you?

**Comment by ranchlai** [comment]
Sorry for late reply and thank you very much for your following-up.
I raise the issue based on the simple thinking:   The model state is most expensive(cost lot of energy) to obtain while the others (e.g., augmentation, scp, spect-augment, scheduler, and even the optim state) are dispensable. In other words, Users should be able to change every other configs  without cause trouble(except model configs of course as it's corrected to model state).

I am not sure if adding more parameters is a good way to achieve this, considering there are already so many of them. 
How do other frameworks (e.g., speechbrain) achieve this?  Maybe not decide now and let's do a litter survey?

**Comment by ranchlai** [comment]
I have checked speechbrain just now.  It seems more flexible for them to decide what to checkpoint and what to load by setting "recoverables", e.g., 
``` yaml
checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
    checkpoints_dir: !ref <save_folder>
    recoverables:
        model: !ref <model>
        scheduler: !ref <lr_annealing>
        counter: !ref <epoch_counter>

log_softmax: !new:speechbrain.nnet.activations.Softmax
    apply_log: True

optimizer: !name:torch.optim.Adam
    lr: !ref <lr>
    betas: (0.9, 0.98)
    eps: 0.000000001
```
They have a checkpoint.py to [https://github.com/speechbrain/speechbrain/blob/9fdce2bbf8129583883769a3728e5964d0b36149/speechbrain/utils/checkpoints.py](url) to do make that possible. 

For us,  can we just detect the diff of hyper-params between loaded state and config file? For example, 
``` python
scheduler = CosineAnnealingWarmRestarts(opti,T_0: 100)
scheduler.state_dict()
>> 
{'T_0': 100,
 'T_i': 100,
 'T_mult': 1,
 'eta_min': 0,
 'base_lrs': [0.1],
 'last_epoch': 0,
 '_step_count': 0,
 'verbose': False,
 'T_cur': 0,
 '_get_lr_called_within_step': False,
 '_last_lr': [0.1]}
```
If users set new T_0 to 200, then the state for scheduler should be discarded, just like if they change scheduler to StepLR, we have no choice but to discard the state dict.

**Comment by kamo-naoyuki** [comment]
> For us, can we just detect the diff of hyper-params between loaded state and config file? For example,

We must keep generality for trainer design. Your idea is actually impossible because It's not guaranteed that the parameter name for `__init__` is as same as the name of state, for example

```
class Schduelr:
  df __init__(self,  param):
      self.param = param
  def state_dict(self):
     return {"PARAM": sself.param}
```

Only we can do is detecting the difference between a new configuration and the previous config file, which is saved at the previous run.

However, I think it's not good idea to detect the dfference and set attributes automatically. Setting attributes is dangerous operation and normally unrecommended in general.

**Comment by kamo-naoyuki** [comment]
> I have checked speechbrain just now. It seems more flexible for them to decide what to checkpoint and what to load by setting "recoverables", e.g.,
> 

Maybe, if removing `scheduler ` from `recoverables`, not only loading but also saving is avoided at the same time.

Anyway, we can also add options to control what state will be saved and loaded, but I think this is not your intention.

I guess you need to modify the lr with keeping the scheduler state.

If it's okay to reset the scheduler state, please use `--init_param` option instead.

**Comment by ranchlai** [comment]
Ok, I understand your concern. It's quite ok for me to use --init_param.

**Comment by ranchlai** [comment]
> > I have checked speechbrain just now. It seems more flexible for them to decide what to checkpoint and what to load by setting "recoverables", e.g.,
> 
> Maybe, if removing `scheduler ` from `recoverables`, not only loading but also saving is avoided at the same time.
> 
> Anyway, we can also add options to control what state will be saved and loaded, but I think this is not your intention.
> 
> I guess you need to modify the lr with keeping the scheduler state.
> 
> If it's okay to reset the scheduler state, please use `--init_param` option instead.
I think recoverables can be be edited before training resumes.

**Comment by ranchlai** [comment]
Thank you very much for thinking about this !
